jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/CallbackCompletionHandler.java

```diff
@@ -0,0 +1,874 @@
+     *
+     * @param listener the event listener
+     */
+    public CallbackCompletionHandler(Listener listener)
+    {
+        this.listener = listener;
+    }
+
+    /**
+     * @return the timeout for the completion of the {@link #handle(Request, Response, Callback)} callback
+     */
+    @ManagedAttribute("The timeout for the completion of the handle() callback")
+    public Duration getHandlerCallbackTimeout()
+    {
+        return handlerCallbackTimeout;
+    }
+
+    public void setHandlerCallbackTimeout(Duration timeout)
+    {
+        this.handlerCallbackTimeout = Objects.requireNonNull(timeout);
+    }
+
+    /**
+     * @return whether the {@link #handle(Request, Response, Callback)} callback is completed
+     * in case the {@link #getHandlerCallbackTimeout()} expires
+     * @see #getHandlerCallbackTimeout()
+     */
+    @ManagedAttribute("Whether the handle() callback is completed in case of timeout")
+    public boolean isCompleteHandlerCallbackAtTimeout()
+    {
+        return completeHandlerCallbackAtTimeout;
+    }
+
+    public void setCompleteHandlerCallbackAtTimeout(boolean completeHandlerCallbackAtTimeout)
+    {
+        this.completeHandlerCallbackAtTimeout = completeHandlerCallbackAtTimeout;
+    }
+
+    /**
+     * @return the timeout for the execution of the demand callback passed to {@link Request#demand(Runnable)}
+     */
+    @ManagedAttribute("The timeout for the execution of the demand callback")
+    public Duration getDemandCallbackTimeout()
+    {
+        return demandCallbackTimeout;
+    }
+
+    public void setDemandCallbackTimeout(Duration timeout)
+    {
+        this.demandCallbackTimeout = Objects.requireNonNull(timeout);
+    }
+
+    /**
+     * @return the timeout for the execution of a {@link Response#write(boolean, ByteBuffer, Callback)} call
+     */
+    @ManagedAttribute("The timeout for the execution of a response write")
+    public Duration getWriteTimeout()
+    {
+        return writeTimeout;
+    }
+
+    public void setWriteTimeout(Duration timeout)
+    {
+        this.writeTimeout = Objects.requireNonNull(timeout);
+    }
+
+    /**
+     * @return the timeout for the execution of the response write callback passed to {@link Response#write(boolean, ByteBuffer, Callback)}
+     */
+    @ManagedAttribute("The timeout for the execution of the response write callback")
+    public Duration getWriteCallbackTimeout()
+    {
+        return writeCallbackTimeout;
+    }
+
+    public void setWriteCallbackTimeout(Duration timeout)
+    {
+        this.writeCallbackTimeout = Objects.requireNonNull(timeout);
+    }
+
+    @Override
+    public boolean handle(Request originalRequest, Response originalResponse, Callback originalCallback) throws Exception
+    {
+        StateInfo stateInfo = new StateInfo(originalRequest);
+        stateInfos.put(originalRequest, stateInfo);
+        Request.addCompletionListener(originalRequest, x -> stateInfos.remove(originalRequest));
```

Comment:
avoid capturing lambda:
```suggestion
        Request.addCompletionListener(originalRequest, stateInfo::onComplete);
```

Reply:
It is capturing `stateInfo` anyway, and the original code is clearer as the line above `put()` and this one `remove()`.